### ISSUE TITLE
Evaluate whether standard library docs should use positional-only notation throughout

### ISSUE THREAD

**Issue by rhettinger** [python/steering-council#12]
A pervasive change is being proposed for the documentation.  It has some value in communicating what were formerly considered implementation details and it has value in more precisely specifying the language.  It also has the effect of making the docs less approachable, creating a mental hiccup every time some reads those docs.  Mariatta frequently gives talks suggesting that the existing docs already suffer usability problems for end-users. This change would make the situation worse.

See the [PR and tracker discussion](https://bugs.python.org/issue37134)

Unlike other proposals, we do have some user testing with this and I think we would be foolish to ignore it.  At least twice a month, I give intermediate/advanced training to people already using Python professionally. The reaction to this syntax is almost universally negative.  Nothing from other languages looks like this.  The notation does add information but typically not the information that people are looking for when they look something up in the docs.  It is more than a minor distraction -- it immediately stops a reader in their tracks to decipher what it means:

```
b2a_uu(data, /, *, backtick=False)
    Convert binary data to a line of ASCII characters, the return value is the
   converted line, including a newline char.

compress(data, /, level=-1)
   Compresses the bytes in *data*, returning a bytes object containing compressed data.
```

I agree that a person can be trained to make sense of this; however, I also think a person shouldn't have to have training to read help(). If they are using help(), they are already alone and are looking for help.

In the cases like ``list(iterable=(), /)``, ``pow(x, y, z=None, /)``, ``sum(iterable, start=0, /)``, I don't imagine that a single user will ever be helped by this.  In this history of Python, no actual user has ever complained that ``len(obj)`` was inadequate and that they would have been better served by ``len(obj, /)``.

I thought this warranted attention by the Steering Council because this changes the face of Python in a somewhat sweeping manner.

For myself, now speaking as user of Python rather than on behalf of users, I would personally get no value out of any of proposed doc changes.

**Comment by brettcannon** [comment]
I've gone ahead and added this to the agenda for our next meeting.

**Comment by willingc** [comment]
Thanks @brettcannon for adding to the agenda and thanks @rhettinger for raising the issue. It would be great to look for solutions that move past "all or nothing" documentation. From a teaching standpoint, I can see value in a softer introduction to users; from an API maintainer, I see value in more robust information. I personally would like to see us work toward documentation which will serve both groups.

**Comment by bskinn** [comment]
I don't know anything about the feasibility of it, but this reminds me of something that was raised at the packaging mini-summit last month, related to whether to emphasize the flat or the src project layout in the beginners' packaging tutorial.  Someone suggested writing the docs such that they could reflect *either* option, toggled at the webpage level at the preference of the reader.

Similarly, would it be possible to write the API docs both without (the default, probably) and with the confusing-to-beginners annotations? I actually just came across the [`sphinx-tabs`](https://github.com/djungelorm/sphinx-tabs) Sphinx extension, that might provide an elegant way to handle it -- in particular the [grouped tabs](https://github.com/djungelorm/sphinx-tabs#grouped-tabs) feature.

**Comment by brettcannon** [comment]
@bskinn won't know until someone tries. 😉 If Sphinx provides a way to hook into the parsing of the directives then it might be possible.

**Comment by bskinn** [comment]
@rhettinger Would it be desirable also to suppress the `*` marker for keyword-only arguments in signatures, in a 'simplified' API docs view?

**Comment by willingc** [comment]
> Someone suggested writing the docs such that they could reflect _either_ option, toggled at the webpage level at the preference of the reader.
> 
> Similarly, would it be possible to write the API docs both without (the default, probably) and with the confusing-to-beginners annotations? I actually just came across the [`sphinx-tabs`](https://github.com/djungelorm/sphinx-tabs) Sphinx extension, that might provide an elegant way to handle it -- in particular the [grouped tabs](https://github.com/djungelorm/sphinx-tabs#grouped-tabs) feature.

That's very similar to my thinking as well. I think that `sphinx-tabs` or something similar might do the trick. If you have time, we would love for you to give it a try :wink:

**Comment by willingc** [comment]
@bskinn Nice to see that you added yourself to the nosy list on the bpo issue.

**Comment by warsaw** [comment]
> In the cases like list(iterable=(), /), pow(x, y, z=None, /), sum(iterable, start=0, /), I don't imagine that a single user will ever be helped by this.

I fully agree.  I can't see any problem with *at least* suppressing a trailing slash.

**Comment by bskinn** [comment]
&lt;grin&gt; I'm still relatively new to core work, but... definitely not my first rodeo on bpo.

Proof-of-concept project in progress here: https://github.com/bskinn/tabslash.

Publicly viewable docs build took a bit of doing, as ReadTheDocs is (naturally) [still working with 3.7](https://readthedocs.org/projects/tabslash/builds/9173601/). I've got my current proof-concept docs build exposed at http://skinn.us.to:11115, though, for people to more easily inspect and comment on. 

The current version of `sphinx.ext.autodoc` parses function definitions with `/` just fine, but doesn't render the slash in the output (generated with `.. autofunction:: tabslash.tabslash.foo`):

![image](https://user-images.githubusercontent.com/11325439/58915484-1ccccb80-86ef-11e9-93f0-81e78641541d.png)

It's easy enough to [manually craft the function signature](https://raw.githubusercontent.com/bskinn/tabslash/07fcebf53f78dd8aef92c046e4dbc18801c0a2c1/doc/source/index.rst) to include the slash, but that's obviously not a scalable solution (generated with `.. autofunction:: tabslash.tabslash.foo(a, b, /)`):

![image](https://user-images.githubusercontent.com/11325439/58915917-2571d180-86f0-11e9-96f1-91a60f6e97a5.png)

As discussed in the bpo thread, ideally we could catch the ear of the Sphinx team before they make any modifications to `autodoc` to add the functionality to display the posarg-only slash, thus allowing the chance to provide input into their process, to make whatever documentation approach is decided upon easier to implement.

**Comment by bskinn** [comment]
Updated to use `python-docs-theme`.

`.. autofunction:: tabslash.tabslash.foo`:

![image](https://user-images.githubusercontent.com/11325439/58916601-1d1a9600-86f2-11e9-857d-77beca1756e3.png)

`.. autofunction:: tabslash.tabslash.foo(a, b, /)`

![image](https://user-images.githubusercontent.com/11325439/58916638-40454580-86f2-11e9-9602-1484e27607b4.png)

**Comment by ncoghlan** [comment]
If we go with the tabbed approach, perhaps "API Essentials" vs "Full API Details"?

**Comment by ncoghlan** [comment]
Adding some notes from the SC discussion at 2019-06-04/5:

* for Python 3.8 specifically, we think it makes sense to continue to leave the slashes out of the documentation for the all parameters are positional-only case (i.e. `... , /)`) and the all parameters are positional-only & keyword-only case (i.e. `..., /, *, ...` and `..., /, *args, ...`). This question can then be revisited for Python 3.9.
* however, we actively want to see them added for the cases where an API has a mixture of positional-only and positional-or-keyword args (i.e. `..., /, ...`)
* we'd like to see the rendered documentation for function signatures enhanced to use tooltips to name the positional-only (bare `/`), keyword-only (bare `*`), additional positional args (named `*`), and additional keyword args (named `**`) notations (hyperlinking to a glossary entry could also be interesting, but may be too visually noisy based on how the hyperlinks get rendered)

The 2-tabs idea is interesting, but needs to be considered carefully in terms of the potential implications for both docs readability, and docs maintainability.

**Comment by ncoghlan** [comment]
Suggestions for notation tooltips (these are directly from me, not from the SC discussion):

* bare `/`: "preceding parameters are positional-only arguments"
* bare `*`: "following parameters are keyword-only arguments"
* `*args`: "collects any additional positional arguments, any following parameters are keyword-only arguments"
* `**kwds`: "collects any additional keyword arguments"

(Note: tooltips are problematic on mobile devices, since "hover" isn't something touch UIs support, but there are also a lot of other mobile usability challenges to be considered with the documentation)

**Comment by bskinn** [comment]
@ncoghlan I don't know offhand how to easily add tooltips to specific character(s) in a Sphinx docset, especially for content in the `{arg}` of, e.g., `.. function:: {arg}` or `.. autofunction:: {arg}`. A plugin may already exist... or, some combination of changes to `python-docs-theme` and/or the Sphinx `python` domain itself may be required.

Since `/`, `*`, `*args` and `**kwargs` are fixed strings, it might be as simple as a search-and-apply, though.

**Comment by bskinn** [comment]
@ncoghlan New version of docs with your suggested tab titles, at http://skinn.us.to:11115/748a243:

![image](https://user-images.githubusercontent.com/11325439/58971385-58b46f00-8789-11e9-8a41-1132704229cf.png)

I'm not particularly fond of how tall the tab bar is.

**Comment by pradyunsg** [comment]
That can be fixed via CSS:

```
.sphinx-tabs .ui .item.container {
    padding-top: 0;
    padding-bottom: 0;
}
```

**Comment by ncoghlan** [comment]
@bskinn My assumption is that the tooltip idea would need to a Sphinx enhancement in the way the `python` domains renders functions and classes. There might be other ways of doing it, but I think they'd end up being hacky workarounds vs doing it right.

**Comment by bskinn** [comment]
@ncoghlan Agreed. It could probably be added to the builtin `python` domain, but configured to recognize a toggle option... `:argument-tooltips:`, say, in the rST, as well as something similar for `conf.py`. That way other projects would have the choice as to whether to display them or not.

`intersphinx` is [already configured](https://sphobjinv.readthedocs.io/en/latest/syntax.html#syntax-mouseover-example) to lay in tool tips like this, so the actual machinery itself should hopefully transfer ~neatly.

**Comment by willingc** [comment]
This appears to be addressed for 3.8. Thanks all for your suggestions and comments. I'm going to close this particular issue. If further specific issues should be open, bpo seems a reasonable place to file.
